cirq/contrib/two_qubit_gates/gate_compilation.py

```diff
@@ -0,0 +1,328 @@
+    spacing = np.sqrt(max_infidelity / 3)
+    mesh_points = weyl_chamber_mesh(spacing)
+
+    num_samples = mesh_points.shape[0] * 50
+
+    # include the base gate itself
+    kak_vecs = [kak_vector(base_gate, check_preconditions=False)]
+    sq_cycles: List[Tuple[_SingleQubitGatePair, ...]] = [()]
+
+    # Tabulate gates that are close to gates in the mesh
+    u_locals_0 = random_qubit_unitary(num_samples)
+    u_locals_1 = random_qubit_unitary(num_samples)
+
+    u_locals_for_gate: Dict[int, Tuple[_SingleQubitGatePair, ...]] = {}
+    tabulation_cutoff = 0.5 * spacing
+    out = _tabulate_KAK_vectors(u_locals_for_gate, base_gate, tabulation_cutoff,
+                                mesh_points, (u_locals_0, u_locals_1))
+    kak_vecs.extend(out[0])
+    sq_cycles.extend(out[1])
+
+    # Will be used later for getting missing KAK vectors.
+    kak_vecs_single = np.array(kak_vecs[1:])
+    sq_cycles_single = list(sq_cycles[1:])
+
+    if verbose:
+        # coverage: ignore
+        print(f'fraction satisfied with 2 gates'
+              f': {len(u_locals_for_gate) / mesh_points.shape[0]:.3f}')
+
+    # repeat for double products
+    # Multiply by the same local unitary!
+    out = _tabulate_KAK_vectors(u_locals_for_gate, base_gate, tabulation_cutoff,
+                                mesh_points, (u_locals_0, u_locals_1),
+                                (u_locals_0, u_locals_1))
+
+    kak_vecs.extend(out[0])
+    sq_cycles.extend(out[1])
+
+    if verbose:
+        # coverage: ignore
+        print(f'fraction satisfied with 2 gates and 3 gates(same single qubit)'
+              f': {len(u_locals_for_gate) / mesh_points.shape[0]:.3f}')
+
+    # If all KAK vectors in the mesh have been tabulated, return.
+    missing_vec_inds = set(range(mesh_points.shape[0])) - set(u_locals_for_gate)
+
+    if not missing_vec_inds:
+        # coverage: ignore
+        return GateTabulation(base_gate, np.array(kak_vecs), sq_cycles,
+                              max_infidelity)
+
+    # Run through remaining KAK vectors that don't have products and try to
+    # correct them
+
+    u_locals_0p = random_qubit_unitary(100)
+    u_locals_1p = random_qubit_unitary(100)
+    u_locals = vector_kron(u_locals_0p, u_locals_1p)
+
+    # Loop through the mesh points that have not yet been tabulated.
+    # Consider their nonlocal parts A and compute products of the form
+    # base_gate^\dagger k A
+    # Compare the KAK vector of any of those products to the already tabulated
+    # KAK vectors from single products of the form
+    # base_gate k0 base_gate.
+    # If they are close, then base_gate^\dagger k A  ~ base_gate k0 base_gate
+    # So we may compute the outer local unitaries kL, kR such that
+    #    base_gate^\dagger k A = kL base_gate k0 base_gate kR
+    #    A = k^\dagger base_gate kL base_gate k0 base_gate kR
+    #    the single-qubit unitary kL is the one we need to get the desired
+    #    KAK vector.
+    base_gate_dag = base_gate.conj().T
```

Comment:
This method is very long. Can it be split into two logical pieces? Maybe the loop body can be extracted into a subroutine?

Reply:
I intend to do this in the future. Currently the algorithm is still under development, as I would like to allow for more than 3 usages of the base gate. I think this refactoring will be more useful once I  think about how to implement this.